### PR TITLE
WFLY-4255 Enable subclassing of Conainer and Client ContainerController-...

### PR DESCRIPTION
**Pull request by rhusar**
Looks like these should be less restrictive..

https://issues.jboss.org/browse/WFLY-4255

### CODE REVIEW

**Comment by smiklosovic** [comment]
From my point of view, making injections protected so you can access them in subclass is considered bad practice. I understand you can have some problem to implement what you want to achieve however doing it protected is IMHO not the answer to that.

**Comment by aslakknutsen** [comment]
@rhusar What does the subclass look like?

**Comment by rhusar** [comment]
The subclass would be something like https://github.com/rhusar/wildfly-arquillian/commit/65c109638416f297b723a4853917f2774e692e6d#diff-2e6c932aa7f23d0c04087183c33d2cffR15

The idea is to extend the existing ContainerController with container-specific lifecycle methods. Take the implementation classes and extend them with the extra method.

It looks like this might not work since I can't override the already registered observer (?) and neither I can't get the classloader right (https://gist.githubusercontent.com/rhusar/91df9453b5a4b8ddbf18/raw/f8949f2bcba341edcf9fc782b058da9ab5a102bc/gistfile1.txt).

Thoughts on whether its doable?

**Comment by aslakknutsen** [comment]
@rhusar For the 'classloading' issue there.. You're registering the client side and the container side on the client side. Which basically duplicates them. The CommandService is only a Container side service, which is why it complains I believe. 

You'll need to implement a AuxilliaryArchiveAppender and create a little jar file that contain; the 'XContainerContainer' class which communicates with the Client side; the creator of the 'XContainerContainer' object; and the RemoteCommand object. Also add a RemoteLoadableExtension to the jar as a SPI file to register the classes in container.

**Comment by aslakknutsen** [comment]
@rhusar Not sure you need to override the observer. No problem having two of them in the context. The'normal' one would be injected via ContainerController.. and you're injected via WildFlyContainerController. Even if ContainerController extend WildFlyContainerController it should work the same. They will provide some 'overlapping' behavior, but that's fine.

File: container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java [hunk 1]
```diff
@@ -48,16 +48,16 @@
...
-   private Event<ContainerControlEvent> event;
+   protected Event<ContainerControlEvent> event;
 
    @Inject
-   private Instance<ContainerRegistry> containerRegistry;
+   protected Instance<ContainerRegistry> containerRegistry;
    
    @Inject
-   private Instance<DeploymentScenario> deploymentScenario;
+   protected Instance<DeploymentScenario> deploymentScenario;
```
Comment (aslakknutsen) [inline container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java]: Replace with getterMethods

File: container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java [hunk 2]
```diff
@@ -48,16 +48,16 @@
 public class ClientContainerController implements ContainerController
 {
 
-   private final Logger log = Logger.getLogger(ClientContainerController.class.getName());
+   protected final Logger log = Logger.getLogger(ClientContainerController.class.getName());
```
Comment (aslakknutsen) [inline container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java]: Create your own logger.. :)

File: container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ContainerContainerController.java [hunk 3]
```diff
@@ -38,7 +38,7 @@
 public class ContainerContainerController implements ContainerController
 {
    @Inject
-   private Instance<ServiceLoader> serviceLoader;
+   protected Instance<ServiceLoader> serviceLoader;
```
Comment (aslakknutsen) [inline container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ContainerContainerController.java]: Replace with protected getterMethod

File: container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java (hunk 2 above)
Comment (rhusar) [inline container/test-impl-base/src/main/java/org/jboss/arquillian/container/test/impl/client/container/ClientContainerController.java]: Haha, yeah, this is wrong, but i do have one in the POC just kept this wrongly...

**Comment by rhusar** [comment]
@aslakknutsen Thanks for the tips, let me see if this works with your suggested changes

**Comment by aslakknutsen** [comment]
@rhusar Did you have any luck here? Planning 1.1.6.Final today, so.. :)

**Comment by rhusar** [comment]
@aslakknutsen I fixed the problems I had in my branch for wildfly-arquillian but still needs some work, but I should be able to update this one -- give me a minute.

**Comment by rhusar** [comment]
Updated.

**Comment by aslakknutsen** [comment]
pushed upstream https://github.com/arquillian/arquillian-core/commit/c178876bbf933c3d226b7282b05f4c8c354f7cd3

**Pull request closed without merging**